store/mockstore/unistore/cophandler/mpp_exec.go

```diff
@@ -0,0 +1,379 @@
+package cophandler
+
+import (
+	"context"
+	"io"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/ngaut/unistore/tikv/dbreader"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/kvproto/pkg/kvrpcpb"
+	"github.com/pingcap/kvproto/pkg/mpp"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/sessionctx/stmtctx"
+	"github.com/pingcap/tidb/store/tikv/tikvrpc"
+	"github.com/pingcap/tidb/tablecodec"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/codec"
+	"github.com/pingcap/tidb/util/rowcodec"
+	"github.com/pingcap/tipb/go-tipb"
+)
+
+// mpp executor that only servers for mpp execution
+type mppExec interface {
+	open() error
+	next() (*chunk.Chunk, error)
+	getFieldTypes() []*types.FieldType
+}
+
+type baseMPPExec struct {
+	sc *stmtctx.StatementContext
+
+	mppCtx *MPPCtx
+
+	children []mppExec
+
+	fieldTypes []*types.FieldType
+}
+
+func (b *baseMPPExec) getFieldTypes() []*types.FieldType {
+	return b.fieldTypes
+}
+
+type tableScanExec struct {
+	baseMPPExec
+
+	kvRanges []kv.KeyRange
+	startTS  uint64
+	dbReader *dbreader.DBReader
+
+	chunks []*chunk.Chunk
+	chkIdx int
+
+	decoder *rowcodec.ChunkDecoder
+}
+
+func (e *tableScanExec) SkipValue() bool { return false }
+
+func (e *tableScanExec) Process(key, value []byte) error {
+	handle, err := tablecodec.DecodeRowKey(key)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	chk := chunk.NewChunkWithCapacity(e.fieldTypes, 0)
+	err = e.decoder.DecodeToChunk(value, handle, chk)
+	e.chunks = append(e.chunks, chk)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+func (e *tableScanExec) open() error {
+	for _, ran := range e.kvRanges {
+		err := e.dbReader.Scan(ran.StartKey, ran.EndKey, math.MaxInt64, e.startTS, e)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+func (e *tableScanExec) next() (*chunk.Chunk, error) {
+	if e.chkIdx < len(e.chunks) {
+		e.chkIdx++
+		return e.chunks[e.chkIdx-1], nil
+	}
+	return nil, nil
+}
+
+type exchSenderExec struct {
+	baseMPPExec
+
+	exchangeSender *tipb.ExchangeSender
+	tunnels        []*ExchangerTunnel
+	outputOffsets  []uint32
+}
+
+func (e *exchSenderExec) open() error {
+	return e.children[0].open()
```

Comment:
return a error when len(e.children[0])!=1?

Reply:
the check of children's number is.... meaningless. Actually the constructing process of children is quite simple: translated from pb. As long as pb has passed one child, there is only one child here too.